### PR TITLE
add func cinder disk enlarge without data damage

### PR DESCRIPTION
**Pull request by jeremyzhou**
@dims @jsafrane 
**What this PR does / why we need it**:
1.add function cinder disk enlarge without data damage,(ext2,ext3,ext4)

**Special notes for your reviewer**:
Function test step:
1.Create cinder PV,PVC,RC
PVC:
{
            "apiVersion": "v1",
            "kind": "PersistentVolumeClaim",
            "metadata": {
                "name": "block1",
                "namespace": "test"
            },
            "spec": {
                "accessModes": [
                    "ReadWriteOnce"
                ],
                "resources": {
                    "requests": {
                        **"storage": "1Gi"**
                    }
                },
                "selector": {
                    "matchLabels": {
                        "tenantId": "test",
                        "volumeName": "block1"
                    }
                },
                "volumeName": "block1pv"
            }
        }
PV:
{
            "apiVersion": "v1",
            "kind": "PersistentVolume",
            "metadata": {
                "labels": {
                    "tenantId": "test",
                    "volumeName": "block1"
                },
                "name": "block1pv",
                "namespace": ""
            },
            "spec": {
                "accessModes": [
                    "ReadWriteOnce"
                ],
                "capacity": {
                    **"storage": "1Gi"**
                },
                "cinder": {
                    "fsType": "ext4",
                    "volumeID": "9a97658b-da74-4b64-8bb5-8901fc738567"
                },
                "persistentVolumeReclaimPolicy": "Retain"
            }
        }
RC:
{
    "apiVersion": "v1",
    "kind": "ReplicationController",
    "metadata": {
        "name": "testrc",
        "namespace": "test"
    },
    "spec": {
        "replicas": 1,
        "selector": {
            "name": "testrc-1"
        },
        "template": {
            "metadata": {
                "annotations": {
                    "namespace": "test",
                    "networks": "{\"ports\": [{\"attach_to_network\": \"lan\", \"attributes\": {\"accelerate\": \"false\", \"function\": \"std\", \"nic_name\": \"eth0\", \"nic_type\": \"normal\"}}]}"
                },
                 "labels": {
                     "name": "testrc-1"
                  }
            },
            "spec": {
                "containers": [
                    {
                        "image": "101.101.101.157:6666/test/projectmana:v1",
                        "imagePullPolicy": "Always",
                        "name": "cradlecontroller",
                        "securityContext": {
                            "privileged": false
                        },
                        "volumeMounts": [
                            {
                                "mountPath": "/mnt",
                                "name": "volume"
                            }
                        ]
                    }
                ],
                "volumes": [
                    {
                        "name": "volume",
                        "persistentVolumeClaim": {
                            "claimName": "block1"
                        }
                    }
                ]
            }
        }
    }
}

2.exec into POD and use df -h check cinder volume size
image

3.Use dd cmd write a file in cinder volume
image

4.delete Pod, and extends cinder volume size to 2Gi when cinder volume status is available
image

5.rebuild POD, and exec into POD, use df -h check cinder volume size,use dd cmd test cinder volume
image

### CODE REVIEW

File: pkg/util/mount/mount_linux.go [hunk 1]
```diff
@@ -492,6 +492,19 @@ func (mounter *SafeFormatAndMount) formatAndMount(source string, target string,
 		}
 	}
 
+	// Run e2fsck and resize2fs to extends ext2/ext3/ext4 disk size
```
Comment (zhangxiaoyu-zidif) [inline pkg/util/mount/mount_linux.go]: if this only works for cinder, you need add some `if` statement or in volume/cinder module.

File: pkg/util/mount/mount_linux.go [hunk 2]
```diff
@@ -492,6 +492,19 @@ func (mounter *SafeFormatAndMount) formatAndMount(source string, target string,
...
+	}
+	args = []string{source}
+	out, err = mounter.Exec.Run("resize2fs", args...)
+	if err != nil {
+		glog.Infof("`resize2fs` error %s", string(out))
+	}
+
 	// Try to mount the disk
 	glog.V(4).Infof("Attempting to mount disk: %s %s %s", fstype, source, target)
 	mountErr := mounter.Interface.Mount(source, target, fstype, options)
```
Comment (zhangxiaoyu-zidif) [inline pkg/util/mount/mount_linux.go]: your code will affect here.

**Comment by zhangxiaoyu-zidif** [comment]
sign CLA first.

File: pkg/util/mount/mount_linux.go [hunk 3]
```diff
@@ -492,6 +492,19 @@ func (mounter *SafeFormatAndMount) formatAndMount(source string, target string,
 		}
 	}
 
+	// Run e2fsck and resize2fs to extends ext2/ext3/ext4 disk size
+	glog.Infof("Extending disk size run cmd e2fsck on disk: %s", source)
+	args = []string{"-f", "-y", source}
+	out, err = mounter.Exec.Run("e2fsck", args...)
+	if err != nil {
+		glog.Infof("`e2fsck` error %s", string(out))
```
[the remark this answers is not in the document]
Comment (jeremyzhou) [reply pkg/util/mount/mount_linux.go]: 1.these cmd will not affect function formatAndMount progress
2.if cinder volume have not  formated, there must be error
so use glog.Infof

File: pkg/util/mount/mount_linux.go (hunk 1 above)
Comment (jeremyzhou) [reply pkg/util/mount/mount_linux.go]: formatAndMount function use by fc,iscsi,photon...

**Comment by zhangxiaoyu-zidif** [comment]
test failed：
```shell
--- FAIL: TestSafeFormatAndMount (0.01s)
	safe_format_and_mount_test.go:193: Unexpected command: e2fsck [-f -y /dev/foo]
	safe_format_and_mount_test.go:193: Unexpected command: resize2fs [/dev/foo]
	safe_format_and_mount_test.go:193: Unexpected command: e2fsck [-f -y /dev/foo]
	safe_format_and_mount_test.go:193: Unexpected command: resize2fs [/dev/foo]
	safe_format_and_mount_test.go:193: Unexpected command: e2fsck [-f -y /dev/foo]
	safe_format_and_mount_test.go:193: Unexpected command: resize2fs [/dev/foo]
	safe_format_and_mount_test.go:199: Unexpected command e2fsck. Expecting lsblk
	safe_format_and_mount_test.go:203: Unexpected args [-f -y /dev/foo]. Expecting [-n -o FSTYPE /dev/foo]
	safe_format_and_mount_test.go:203: Unexpected args [-f -y /dev/foo]. Expecting [-n -o FSTYPE /dev/foo]
	safe_format_and_mount_test.go:203: Unexpected args [-f -y /dev/foo]. Expecting [-n -o FSTYPE /dev/foo]
	safe_format_and_mount_test.go:193: Unexpected command: resize2fs [/dev/foo]
	safe_format_and_mount_test.go:193: Unexpected command: lsblk [-n -o FSTYPE /dev/foo]
	safe_format_and_mount_test.go:193: Unexpected command: mkfs.ext4 [-F /dev/foo]
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range
```

**Comment by gnufied** [comment]
We already support resizing cinder volumes.  https://kubernetes.io/docs/concepts/storage/persistent-volumes/#expanding-persistent-volumes-claims is documentation of expanding persistent volumes. In 1.9 - this is a alpha feature and requires enabling the corresponding feature gate.

**Comment by gnufied** [comment]
I am going to close this PR. If you run into problems trying to expand cinder volume using 1.9 feature, please open a bug report.

/close

**Pull request closed without merging**